packages/ai-llamafile/src/browser/llamafile-frontend-application-contribution.ts

```diff
@@ -0,0 +1,63 @@
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { LanguageModelRegistry } from '@theia/ai-core';
+import { FrontendApplicationContribution, PreferenceService } from '@theia/core/lib/browser';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { LlamafileLanguageModel } from '../common/llamafile-language-model';
+import { LlamafileServerManager } from '../common/llamafile-server-manager';
+import { LlamafileEntry, PREFERENCE_LLAMAFILE } from './llamafile-preferences';
+
+@injectable()
+export class LlamafileFrontendApplicationContribution implements FrontendApplicationContribution {
+
+    @inject(PreferenceService)
+    protected preferenceService: PreferenceService;
+
+    @inject(LanguageModelRegistry)
+    protected languageModelRegistry: LanguageModelRegistry;
+
+    @inject(LlamafileServerManager)
+    protected llamafileServerManager: LlamafileServerManager;
+
+    onStart(): void {
+        this.preferenceService.ready.then(() => {
+            const llamafiles = this.preferenceService.get<LlamafileEntry[]>(PREFERENCE_LLAMAFILE, []);
+
+            const models = llamafiles.map(llamafile =>
+                LlamafileLanguageModel.createNewLlamafileLanguageModel(llamafile.name, llamafile.uri, llamafile.port, this.llamafileServerManager));
+            this.languageModelRegistry.addLanguageModels(models);
+
+            this.preferenceService.onPreferenceChanged(event => {
+                if (event.preferenceName === PREFERENCE_LLAMAFILE) {
+                    // old models in the preference
+                    const oldModels = new Set((event.oldValue as LlamafileEntry[]).map(v => v.name));
+                    // new models in the preference as map to if a model was added and be able to create the llm and register it
+                    const newModels = (event.newValue as LlamafileEntry[]).reduce((acc, v) => { acc.set(v.name, v); return acc; }, new Map<string, LlamafileEntry>());
+
+                    const modelsToRemove = [...oldModels.values()].filter(model => !newModels.has(model));
+                    const modelDescriptionsToAdd = [...newModels.values()].filter(model => !oldModels.has(model.name));
+
+                    this.languageModelRegistry.removeLanguageModels(modelsToRemove);
+
+                    const modelsToAdd = modelDescriptionsToAdd.map(llamafile =>
+                        LlamafileLanguageModel.createNewLlamafileLanguageModel(llamafile.name, llamafile.uri, llamafile.port, this.llamafileServerManager));
+                    this.languageModelRegistry.addLanguageModels(modelsToAdd);
```

Comment:
We directly add the language models in the frontend here. We should not do that as this will not work in a Cloud deployment, e.g. in "Try Online" without opening up additional ports. As the ports are even configurable, these is even harder.

Instead we should manage the language model in the backend as we do with ollama and openai.